feat: RUM support for Fragments

miquelbeltran opened the pull request:
## feat: RUM support for Fragments

### Description :memo:

This PR adds the RUM functionality to the Android Fragment lifecycle.

The `FragmentManager` gets registered automatically when using `AppCompatActivity` in apps, as RUM will obtain the `FragmentManager` from it and listen to its lifecycle changes.

There are some cases not covered:

- Apps using legacy `FragmentManager` and not the `androidx` support one.
- Apps that manually create their `FragmentManager` instead of using the one in `AppCompatActivity`.
- Any kind of non-Fragment based navigation.

For these cases, we can add ways to register the `FragmentManager` manually (both the support and the legacy ones), as well as ways to register screen navigation events manually. We can do that in a further future PR.

**Type of change**

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

We probably need to mention in the docs that it requires the Activity to be a `AppCompatActivity` and use the support `FragmentManager` _as they should be doing already_ :wink: 

**Updates**

- Split `RUM` into `RUM` and `RUMActivity`
- Moved RUM code to subpackage `.rum`.
- Created `RUMFragment`
- Change example to use Fragments
- Added unit tests for `RUMFragment` and `RUMActivity`

### Screenshots :camera:

<details>
  <summary>Raygun.com Screenshots</summary>

![Screenshot from 2025-01-08 11-37-06](https://github.com/user-attachments/assets/5c5297f5-73eb-4b2d-b641-c188aa1b4e45)
![Screenshot from 2025-01-08 11-36-47](https://github.com/user-attachments/assets/133e7380-3892-4896-bb1a-757140f38653)

</details>

<details>
  <summary>App Screenshots</summary>

![Screenshot_20250108_113346](https://github.com/user-attachments/assets/4f629e29-aefc-41c6-8beb-9ee7fbe333ec)
![Screenshot_20250108_113325](https://github.com/user-attachments/assets/a4b23a95-dc2c-40f1-b4e9-455a528085f1)

</details>

### Test plan :test_tube:

- Tested on sample app
- Added unit tests for `RUMFragment` and `RUMActivity`

### Author to check :eyeglasses:
- [x] Project and all contained modules build
- [x] Tested on appropriate devices/emulators and SDK versions
- [x] Reviewed by another developer
- [ ] Appropriate documentation written (code comments, internal docs)

### Reviewer to check :heavy_check_mark:
- [ ] Change has been tested on a device/emulator
- [ ] Code is written to standards
- [ ] Appropriate tests have been written (code comments, internal docs)

miquelbeltran committed:
WIP attach RUM to Fragment Manager

miquelbeltran committed:
logs similar to Activity rum

miquelbeltran committed:
WIP RUM split in three classes

miquelbeltran committed:
basic RUM Fragment implementation

miquelbeltran committed:
detach from MainActivity FragmentManager

miquelbeltran committed:
WIP fragment loaded event

miquelbeltran committed:
WIP fragments

TheRealAgentK commented on provider/src/main/java/com/raygun/raygun4android/RaygunClient.java:
```diff
@@ -7,10 +7,14 @@
 import android.content.pm.PackageManager;
 import android.os.Bundle;
 
+import androidx.fragment.app.FragmentActivity;
```
Do we need the references to the FragmentAct and Manager in `RaygunClient`?

TheRealAgentK commented:
I think the general approach of splitting `RUM` into `RUMActivity` and `RUMFragment` is the right one. Left one comment re the Fragment references in `RaygunClient` - that feels a bit "hmmmmmm....". But we also import `Activity` and `Application` in there, so, take that comment with a bucket of salt.

TheRealAgentK commented:
Just for reference: @miquelbeltran and I are having additional discussions around how to do this the best/cleanest way outside of the draft PR.

miquelbeltran commented:
> Left one comment re the Fragment references in RaygunClient - that feels a bit "hmmmmmm....". But we also import Activity and Application in there, so, take that comment with a bucket of salt.

Did you post the comment? I cannot find it

miquelbeltran committed:
remove option to manually register a fragment manager, support only androidx fragment activities

miquelbeltran committed:
refactor, wip tests

miquelbeltran committed:
add RUMFragmentTest

miquelbeltran committed:
add RUMActivityTests

miquelbeltran commented on provider/src/main/java/com/raygun/raygun4android/RaygunSettings.java:
```diff
@@ -62,7 +62,7 @@ public static HashSet<String> getIgnoredURLs() {
     }
 
-    static HashSet<String> getIgnoredViews() {
+    public static HashSet<String> getIgnoredViews() {
```
Since the `RUM` class is moved to the `.rum` package, it cannot access `getIgnoredViews` per package visibility anymore.

miquelbeltran commented on provider/src/main/java/com/raygun/raygun4android/rum/RUM.java:
```diff
@@ -19,225 +19,136 @@
+    /**
+     * Gets the singleton instance of the RUM class.
+     */
+    public static RUM getInstance() {
```
`RUM` no longer uses static methods, instead access is done with `getInstance()`. This does not affect end users, but internally allows mocking the RUM instance. For example, in `RUMFragmentTest` the `RUM` instance is mocked.

TheRealAgentK replied:
We should do a similar refactoring for `CrashReporting`.

miquelbeltran commented on provider/src/main/java/com/raygun/raygun4android/rum/RUM.java:
```diff
@@ -19,225 +19,136 @@
+    private long lastSeenTime;
+    private String sessionId;
+    private RaygunUserInfo currentSessionUser;
+    private final RUMActivity rumActivity;
```
All the "Activity" logic was moved to `RUMActivity`, which also facilitates testing it in `RUMActivityTest`

miquelbeltran commented on provider/src/main/java/com/raygun/raygun4android/rum/RUM.java:
```diff
@@ -331,9 +242,10 @@ public static void sendRUMTimingEvent(
                     : RaygunClient.getUser();
 
+            // TODO(Miguel Beltran): type = p, what does it mean?
             RaygunRUMTimingMessage timingMessage =
                 new RaygunRUMTimingMessage.Builder(
-                    eventType == RaygunRUMEventType.ACTIVITY_LOADED ? "p" : "n")
+                    eventType == RaygunRUMEventType.ACTIVITY_LOADED || eventType == RaygunRUMEventType.FRAGMENT_LOADED ? "p" : "n")
                     .duration(milliseconds)
                     .build();
```
From my understanding, `p` is used for any screen view, so we should use `p` for both loading Activities and Fragments.

miquelbeltran committed:
add end-of-file lines

miquelbeltran committed:
small cleanup

miquelbeltran committed:
cleanup sample app

miquelbeltran commented:
@TheRealAgentK I did a bit of cleanup and completed the PR description, added screenshots as well to facilitate review. I listed some cases that are not yet covered but I think we can work on those in a new PR.

miquelbeltran committed:
remove comment

miquelbeltran commented:
I will fix the conflicts

miquelbeltran committed:
spotlessApply

miquelbeltran committed:
spotlessApply

miquelbeltran commented:
Rebased to `develop`

TheRealAgentK committed:
Merge branch 'develop' into mb-rum-fragments

# Conflicts:
#	app/src/main/AndroidManifest.xml
#	provider/src/main/java/com/raygun/raygun4android/CrashReporting.java
#	provider/src/main/java/com/raygun/raygun4android/rum/RUM.java

TheRealAgentK committed:
Neutralised API Key :)

TheRealAgentK commented:
Merged `develop` in to solve the conflicts from the changes in #133 and #82 

Edit: Oh, sorry, just realised that you did rebases before in the branch :-/

TheRealAgentK commented on app/build.gradle:
```diff
@@ -142,6 +142,7 @@ dependencies {
     implementation(libs.androidx.constraintlayout.constraintlayout)
     implementation(libs.android.material.material)
     implementation(libs.androidx.multidex.multidex)
+    implementation libs.androidx.activity
```
which syntax `(..)` or not is actually preferable and better?

miquelbeltran replied:
:thinking: both are supported, but I will change it to `()` to keep it the same

TheRealAgentK commented on app/src/main/AndroidManifest.xml:
```diff
@@ -10,21 +10,25 @@
-        </activity>
+            android:parentActivityName=".MainActivity" />
+        <activity
+            android:name=".fragments.NavigationActivity"
```
Aaactually - I think it'd be good to have two additional Activities in the sample.

1. Nav Act as now - hosting one single fragment
2. Another Nav Act - that where fragments can be swapped between so that people see what such a setup will result in the tracking.

miquelbeltran replied:
Implemented that, see my last comment in the PR

TheRealAgentK commented on app/src/main/java/com/raygun/raygun4android/sample/MainActivity.kt:
```diff
@@ -39,6 +40,7 @@ class MainActivity : AppCompatActivity() {
         val textViewAppVersion = findViewById<TextView>(R.id.textView_appVersion)
         val textViewProviderVersion = findViewById<TextView>(R.id.textView_providerVersion)
         val buttonSecondActivity = findViewById<Button>(R.id.button_secondActivity)
+        val buttonNavigationActivity = findViewById<Button>(R.id.button_navigationActivity)
```
See comment around additional activity in manifest - not gonna mention this idea everywhere.

TheRealAgentK commented on build.gradle:
```diff
@@ -12,6 +12,7 @@ buildscript {
 
 plugins {
     alias(libs.plugins.spotless)
+    alias(libs.plugins.kotlin.android) apply false
```
Ah, that's a good change (->.toml)

TheRealAgentK reviewed (approved):
I think it looks overall really good. 

Left a comment for potential sample improvement.

TheRealAgentK commented:
Approved so that you're not blocked and I think it's genuinely the correct approach and solution.

Apart from the comment on extending the sample - this PR might need some documentation changes and additions as well.

miquelbeltran committed:
use parenthesis

miquelbeltran committed:
format

miquelbeltran commented:
I've added a navigation "swap" activity that allows you to test switching between two Fragments using the FragmentManager `replace()` method. During my tests I noticed an error, that the two Fragments shared the same ID, because `fragment.getId()` returned the id of the parent container, not a unique Fragment ID. I changed that to use `fragment.hashcode()` as it is unique per instance, and tested this on Raygun.

![image](https://github.com/user-attachments/assets/bb66beb1-9f93-4117-8dd5-15de9ae3c602)
(slow time reports are due to attaching a debugger and stopping)

Please take a look at the last commit for the changes.

miquelbeltran committed:
remove template string

TheRealAgentK reviewed (approved):
LGTM

TheRealAgentK commented:
LGTM, I'll go through the README later and might add some commentary around the fragment support before I merge it.

TheRealAgentK committed:
Mini doc update

TheRealAgentK committed:
Merge remote-tracking branch 'origin/mb-rum-fragments' into mb-rum-fragments

TheRealAgentK commented:
> LGTM, I'll go through the README later and might add some commentary around the fragment support before I merge it.

Decided to make rewriting the docs a proper GH issue, it doesn't really cover RUM properly at all and needs restructuring and some brushing up.
